docs/design.md

```diff
@@ -354,13 +364,52 @@ Sleeper allows Athena queries to be run over Sleeper data. This is enabled by th
+the rows with the same key will be processed at the same time.
+
+There are two different ways of configuring this logic, depending on whether compactions are being run using
+Java or DataFusion. The table property `sleeper.table.compaction.method` controls which is used - the possible
+values are `java` and `datafusion`.
+
+If Java is being used then an iterator can be applied. An iterator is a function that takes as input a
+`CloseableIterator<Record>` and returns a `CloseableIterator<Record>`. Examples of iterators that perform
+aggregation or filtering can be found in the `example-iterators` module. The iterator should respect the general
+constraints of a compaction: there could be many hundreds of millions of rows processed by a single compaction job,
+so there should be no attempt to buffer lots of rows in memory; there is no guarantee of the order the files in a
+partition will be compacted, or that all of them will be compacted at the same time so the logic should be
+commutative and associative; the output should be sorted by key so in general the row and sort keys should not be
+changed by the iterator.
+
+If one of the fields in a table is a byte array then that could be a serialised version of an arbitrary Java object.
+This allows aggregation of fields that contain complex values, e.g. Accumulo's iterators in
+[Gaffer](https://github.com/gchq/Gaffer) are used to maintain HyperLogLog sketches which are used to quickly
+approximate the degree of a vertex.
+
+If DataFusion is used for compactions then the aggregation and filtering logic can be applied using a table
+property. This is much simpler than writing an iterator, and DataFusion-based compactions that apply filtering and
+aggregation are much more efficient than Java-based compactions with iterators. The only down-side to the
+DataFusion-based compactions is that it is not possible to write arbitrary logic. The table property
+`sleeper.table.iterator.class.name` can be set to `AGGREGATORS` to specify that aggregation and filtering will
+be performed using DataFusion. The actual aggegation and filtering logic is specified in the
+`sleeper.table.iterator.config` property. The format of this will change. Currently to perform aggregation of
+a count field and to take the maximum of a last_seen field would require the property to be `;,sum(count),max(last_seen)`.
+See https://github.com/gchq/sleeper/issues/4344 for more details and see https://github.com/gchq/sleeper/issues/5102
+for the issue to improve this specification.
+
+Note that if DataFusion-based aggregation is specified then the same logic will be applied in a query-time Java
+iterator so that the user sees the correct results. In time this will be replaced by queries executed via DataFusion.
 
 We are in the process of designing a replacement for this that will work with DataFusion rather than just in Java. See
 the following epic: https://github.com/gchq/sleeper/issues/5123
```

Comment:
It looks like this duplicates the explanation two paragraphs above?

Reply:
Fixed